### PR TITLE
more complete ambiguity detection

### PR DESCRIPTION
**Pull request by vtjnash**
previously we stopped checking for ambiguities as soon as we reached the definition. this assumed that jl_args_morespecific was transitive, but in actuality it is only a partial sort and we build the typemap assuming no ambiguities, which sometimes made it even less accurate at finding them.

~~note that I needed to partially revert to printing ambiguities or this doesn't make it very far in bootstrap before throwing a method ambiguity error. the RFH (request for help) is to crowd-source resolving these ambiguities (or determining which aren't applicable due to that argument combination not being meaningful).~~

ref #16252
ref https://github.com/JuliaLang/julia/pull/16220#discussion_r62260888

~~edit: additionally, due to some recent change, we seem to be unable to detect the ambiguities with Varargs previously computed by https://travis-ci.org/JuliaLang/julia/jobs/126188063~~
edit2: I mistake in that PR was making it possible to compute that ambiguity (while breaking everything else), it's not a recent regression to be unable to compute that.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Given the current state of subtyping, ambiguity detection isn't going to be perfect. I don't think it's good to sometimes warn at definition time and sometimes not. That just adds indecision and unclarity on top of imperfection.

It seems to me specificity should still be transitive. Partial orders still obey transitivity.

**Comment by vtjnash** [comment]
> Given the current state of subtyping, ambiguity detection isn't going to be perfect.

So you've mentioned, but also doesn't come up here

> I don't think it's good to sometimes warn at definition time and sometimes not. That just adds indecision and unclarity on top of imperfection.

Sounds great. When did I contradict this? If they don't get printed at definition time in the WIP, bootstrap fails without a warning for this WIP. But good luck working out why ```LoadError(at "sysimg.jl" line 150: LoadError(at "libuv.jl" line 5: LoadError(at "uv_constants.jl" line 4: Base.MethodError(f=Base.#similar(), args=(Array{Symbol, 1}[:UV_UNKNOWN_HANDLE], Symbol, (21,))))))``` happened.

> It seems to me specificity should still be transitive. Partial orders still obey transitivity.

possibly not the exact right concept, but this is the closest description I could muster to the problem evidenced by your example in https://github.com/JuliaLang/julia/pull/16220#discussion_r62260888. I think an unambiguous partial order obeys transitivity. However, the detection here is for whether the ordering is ambiguous (by verifying that the partial order doesn't require an extra intersection signature method).

**Comment by mbauman** [comment]
Is this what you need, Jameson?

```jl
baremodule Mod
const Base = Main.Base
for n in Base.names(Base, true, true)
   Base.isdefined(Base, n) || continue
   try
       for m in Base.methods(Base.getfield(Base, n))
           eval(Mod, :(($n)($(Base.map(x->:(::$x), Base.getindex(m.sig.parameters,Base.colon(2, Base.length(m.sig.parameters))))...)) = 1))
       end
   catch
       Base.println("skipping $n")
   end
end
end

julia> Test.detect_ambiguities(Mod)
74-element Array{Tuple{Method,Method},1}:
# …
```

[full list](https://gist.github.com/mbauman/fadf228e6e13fc022253c7cca2895c65)

**Comment by vtjnash** [comment]
yep, that looks like the same list (https://github.com/JuliaLang/julia/pull/16255/files#diff-080ddbc6b9941ca45cdaa2622943f143R1)

File: base/array.jl [hunk 1]
```diff
@@ -117,7 +117,7 @@ end
 
 ## Constructors ##
 
-similar(a::Array, T, dims::Dims)      = Array(T, dims)
+similar(a::Array, T::Type, dims::Dims)      = Array(T, dims)
```
Comment (tkelman) [inline base/array.jl]: spacing before the =

File: base/rational.jl [hunk 2]
```diff
@@ -340,6 +340,9 @@ function round{T}(::Type{T}, x::Rational{Bool})
     convert(T, x)
 end
 
+round{T}(::Type{T}, x::Rational{Bool}, ::RoundingMode{:Nearest}) = round(T, x)
+round{T}(::Type{T}, x::Rational{Bool}, ::RoundingMode{:NearestTiesAway}) = round(T, x)
+round{T}(::Type{T}, x::Rational{Bool}, ::RoundingMode{:NearestTiesUp}) = round(T, x)
```
Comment (tkelman) [inline base/rational.jl]: why are these needed?

Comment (vtjnash) [inline base/rational.jl]: they resolve the ambiguity of the method that follows

Comment (tkelman) [inline base/rational.jl]: what's the ambiguity though, aren't these 3 subtypes of the 4th?

File: src/gf.c [hunk 3]
```diff
@@ -744,36 +744,40 @@ struct ambiguous_matches_env {
     struct typemap_intersection_env match;
     union jl_typemap_t defs;
     jl_typemap_entry_t *newentry;
+    jl_array_t *shadowed;
+    int after;
 };
+const int eager_ambiguity_printing = 0;
```
Comment (timholy) [inline src/gf.c]: Do we now reliably pick up ambiguous methods `Core`? If not, would it help to turn this on during Travis builds?

**Comment by timholy** [comment]
Do you need help anymore fixing the ambiguities? I just checked out this branch and built with `eager_ambiguity_printing = 1`; didn't see any warnings.

If you already got them all, thanks! And sorry to be late to the party.

Mergeable, right?

**Comment by vtjnash** [comment]
Thanks for the offer. It turned out most of them were simple method typos (repeats of #16254 for other AbstractArrays), not anything complicated so they're all fixed now. This is simply waiting for someone to hit the merge button.

**Comment by timholy** [comment]
> This is simply waiting for someone to hit the merge button.

Well, I can handle that. Thanks for doing this!
